### PR TITLE
CASEFLOW-1264 | Permissions on substitute appellant button

### PR DESCRIPTION
**Pull request by n1zyy**
### Description
Implements [CASEFLOW-1264](https://vajira.max.gov/browse/CASEFLOW-1264), adding permissions to who can see the Substitute Appellant button.

### Testing Plan

We're going to want a few appeals to test different cases. Pull up a Rails terminal (`make c`) and run the following. We'll just need their uuids.

```ruby
# In case you don't just instinctively do this anyway...
RequestStore[:current_user] = User.system_user

# This should be visible to admins only
appeal1 = FactoryBot.create(:appeal, :dispatched_with_decision_issue,
               docket_type: "evidence_submission",
               stream_type: "original",
               disposition: "allowed")

# No one should see an Add Substitute button on this because docket_type="hearing"
appeal2 = FactoryBot.create(:appeal, :dispatched_with_decision_issue,
               docket_type: "hearing",
               stream_type: "original",
               disposition: "allowed")


# This should be visible to COB admins *and* non-admins
appeal3 = FactoryBot.create(:appeal, :dispatched_with_decision_issue,
               docket_type: "evidence_submission",
               stream_type: "original",
               disposition: "dismissed_death")

Then, just for your reference, note the UUIDs:
[appeal1.uuid, appeal2.uuid, appeal3.uuid]
```

1. This includes a seed data update. I would encourage you to reseed your database as part of a full test.
2. Confirm, at `/test/users`, that you have the `recognized_granted_substitution_after_dd` feature flag enabled. (Enable it if it's not already on.)
3. If you haven't already, create the 3 appeals above.
4. Open 3 tabs in your browser, one for each appeal (at `/queue/appeals/APPEAL_UUID`) as we'll be referencing them several tmes.
5. Log in as BVAAABSHIRE (or any other non-COB user). There should not be an "+ Add Substitute" button on any of them.
6. Log in as COB_USER, which is a non-admin Clerk of the Board user. Only appeal3 should show an "+ Add Substitute" button, because cases with the "dismissed_death" disposition allow all COB users to add a substitute appellant.
7. Log in as BVATCOBB, an admin Clerk of the Board user. appeal1 and appeal3 should show "+ Add Substitute", because admins are able to add substitute appellants to cases regardless of disposition.
8. appeal2 should never show "+ Add Disposition" in any of the examples, because it has a docket type of hearing.
9. Also review the feature test I added and make sure you agree that it covers the somewhat complex acceptance criteria.

### User Facing Changes
Unless anyone objects, I'm going to omit screenshots. I don't think they would be useful here&mdash;the button is not new, and we're just changing the logic of when it's displayed. That would not be clear from the screenshots.

### Storybook Story
I am having trouble with Storybook locally, but it should be unchanged.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -317,14 +317,21 @@ def user_info_for_idt
...
 
   def selectable_organizations
-    orgs = organizations.select(&:selectable_in_queue?)
+    orgs = organizations.select(&:selectable_in_queue?).map do |org|
+      {
+        name: org.name,
+        url: org.url,
+        is_admin: org.admins.include?(self)
+      }
+    end
```
Comment (n1zyy) [inline app/models/user.rb:326]: We want to back these changes out in the morning. The business below with judge teams turns out to make this inaccurate, and ironing this out is just too much of a distraction for this PR.

File: client/app/queue/CaseDetailsView.jsx [hunk 2]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
     [appeal, tasks]
   );
 
-  const appealIsDispached = appeal.status === 'dispatched';
+  const appealIsDispatched = appeal.status === 'dispatched';
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:131]: Unrelated change: fix typo in variable name.

File: client/app/queue/CaseDetailsView.jsx [hunk 3]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
+  // I think we want this instead, but it's kind of off-topic here:
+  // const appealIsDispatched = ['dispatched', 'post_dispatch'].includes(appeal.status);
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:133]: This came up on Bat Team with a user who couldn't add remands to an appeal that was in post-dispatch state. I think this commented-out change would resolve the issue, but I'm not sure we have certainty on whether we want it.

File: client/app/queue/CaseDetailsView.jsx [hunk 4]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
+
+  // Is CAVC remand supported?
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:135]: This can go; was mentally useful as I split stuff apart.

File: client/app/queue/CaseDetailsView.jsx [hunk 5]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
   const supportCavcRemand =
     currentUserIsOnCavcLitSupport && props.featureToggles.cavc_remand;
+
+  // We still need to check whether appeal.decisionIssues includes any
+  // where disposition==='death_dismissal'
+  // And, you know, tests.
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:141]: TODO in the morning.

File: client/app/queue/CaseDetailsView.jsx [hunk 6]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
   const supportSubstituteAppellant =
     currentUserOnClerkOfTheBoard &&
-    props.featureToggles.recognized_granted_substitution_after_dd;
+    props.featureToggles.recognized_granted_substitution_after_dd &&
+    appeal.caseType === 'Original' && // original, vacate, de_novo, court_remand
+    appeal.docketType !== 'hearing' &&
+    appeal.decisionIssuesStatus;
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:147]: Whoops, this last one is totally wrong. This was the start of looking at `appeal.decisionIssues` but somehow I got something else here.

I got hung up trying to get a relevant decision issue added to my appeal on the backend so I could make sure I understood what the data actually looks like.

File: client/app/queue/CaseDetailsView.jsx [hunk 7]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
+
+  // Should we show a post-dispatch box?
+  // Both CAVC remands and substitute appellants are AMA-only.
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:150]: Also comments that should probably be ripped out.

File: client/app/queue/CaseDetailsView.jsx [hunk 8]
```diff
@@ -128,14 +128,29 @@ export const CaseDetailsView = (props) => {
...
   const showPostDispatch =
-    appealIsDispached && (supportCavcRemand || supportSubstituteAppellant);
+    appealIsDispatched && (supportCavcRemand || supportSubstituteAppellant) &&
+      !appeal.isLegacyAppeal;
```
Comment (n1zyy) [inline client/app/queue/CaseDetailsView.jsx:153]: This is debatable. Adding CAVC remands and substitute appellants are both only available on AMA appeals, so I checked that here.  Reflecting on it, I think it's probably more correct to have it in each of the above, in case someone reuses `supportSubstituteAppellant` or `supportCavcRemand` elsewhere.

File: db/seeds/users.rb [hunk 9]
```diff
@@ -356,9 +356,14 @@ def create_clerk_of_the_board_users
         full_name: "Clark ClerkOfTheBoardUser Bard"
       )
       ClerkOfTheBoard.singleton.add_user(atty)
+
```
Comment (n1zyy) [inline db/seeds/users.rb:359]: Break apart the creation of attorney, judge, and admin.

File: db/seeds/users.rb [hunk 10]
```diff
@@ -356,9 +356,14 @@ def create_clerk_of_the_board_users
...
       judge = create(:user, full_name: "Judith COTB Judge", css_id: "BVACOTBJUDGE")
       create(:staff, :judge_role, sdomainid: judge.css_id)
       ClerkOfTheBoard.singleton.add_user(judge)
+
+      admin = create(:user, full_name: "Ty ClerkOfTheBoardAdmin Cobb", css_id: "BVATCOBB")
+      ClerkOfTheBoard.singleton.add_user(admin)
+      OrganizationsUser.make_user_admin(admin, ClerkOfTheBoard.singleton)
```
Comment (n1zyy) [inline db/seeds/users.rb:366]: We had no COB admin with which to test, so I added [The Georgia Peach](https://en.wikipedia.org/wiki/Ty_Cobb), a baseball player and Army captain in WW1. I totally knew about that and didn't find this while trying to find a test-user name to rival the imagination of Clark Bard above. ;)

File: spec/feature/queue/substitute_appellant/shared_setup.rb [hunk 11]
```diff
@@ -35,7 +35,7 @@
 end
 
 RSpec.shared_examples("fill substitution form") do
-  it "allows user to designate a subsitute appellant" do
+  it "allows user to designate a substitute appellant" do
```
Comment (n1zyy) [inline spec/feature/queue/substitute_appellant/shared_setup.rb:38]: Ended up not using anything in here, but fixed this while I was there anyway.

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -366,7 +366,8 @@
       it "returns assigned cases link for judge" do
         is_expected.to include(
           name: "Assign #{user.css_id}",
-          url: format("/queue/%<id>s/assign", id: user.css_id)
+          url: format("/queue/%<id>s/assign", id: user.css_id),
+          is_admin: true
```
Comment (n1zyy) [inline spec/models/user_spec.rb:370]: Back these changes out.

File: client/app/queue/CaseDetailsView.jsx (hunk 3 above)
Comment (n1zyy) [reply client/app/queue/CaseDetailsView.jsx:136]: Want me to pull this out into a separate PR?

File: spec/feature/queue/case_details_spec.rb [hunk 13]
```diff
@@ -1781,6 +1781,91 @@
...
+          # whether the content is present or not!
+          find("div", id: "caseTitleDetailsSubheader")
+          expect(page.has_no_content?(COPY::SUBSTITUTE_APPELLANT_BUTTON)).to eq(true)
+        end
+      end
+
+      shared_examples "the button is shown" do
+        it "the 'Add Substitute' button is shown" do
+          visit "/queue/appeals/#{appeal.external_id}"
+          find("div", id: "caseTitleDetailsSubheader")
```
Comment (n1zyy) [inline spec/feature/queue/case_details_spec.rb:1818]: This is here for the same reason as the one on 1810, but I worry the intent isn't clear.

I'm of half a mind to create a method named something like `wait_for_page_to_render` that just contains this find and a comment, to make it more obvious what's happening. Would that be clearer or weird?

File: spec/feature/queue/case_details_spec.rb [hunk 14]
```diff
@@ -1781,6 +1781,91 @@
...
+
+        it_behaves_like "the button is not shown"
+      end
+
+      context "when the docket type is 'hearing'" do
+        let(:docket_type) { "hearing" }
+
+        context "when the user is an admin" do
+          before { OrganizationsUser.make_user_admin(cob_user, ClerkOfTheBoard.singleton) }
+          after { OrganizationsUser.remove_admin_rights_from_user(cob_user, ClerkOfTheBoard.singleton) }
```
Comment (jcq) [inline spec/feature/queue/case_details_spec.rb:1840]: TIL about these org admin methods

File: spec/feature/queue/case_details_spec.rb [hunk 15]
```diff
@@ -1781,6 +1781,91 @@
...
+        FeatureToggle.enable!(:recognized_granted_substitution_after_dd)
+      end
+
+      context "When the case type is not 'original'" do
+        let(:case_type) { "de_novo" }
+
+        it_behaves_like "the button is not shown"
+      end
+
+      context "when the docket type is 'hearing'" do
```
Comment (jcq) [inline spec/feature/queue/case_details_spec.rb:1835]: it sorta feels like there should be a context for "when docket type is not hearing" at same level, but in theory we'll be able to remove this section soon when we implement the stuff for hearing docket

**Review by jcq — APPROVED** [review]
This looks like it meets all the AC. Good work on fighting with the tests for the (non)appearance of the button.

File: spec/feature/queue/case_details_spec.rb (hunk 15 above)
Comment (n1zyy) [reply spec/feature/queue/case_details_spec.rb:1835]: I struggled with this a bunch: I didn't really want to test all the permutations (feature flag enabled? × case type original? × docket type of hearing? × disposition is "dismissed_death" ...), so I settled for the base case in `let(:appeal)...` and then modified one setting at a time in each context.

It'll be nice to see this one go, though.

File: spec/feature/queue/case_details_spec.rb (hunk 14 above)
Comment (n1zyy) [reply spec/feature/queue/case_details_spec.rb:1840]: I sort of feel like it'd be cleaner if we could write `ClerkOfTheBoard.singleton.make_user_admin(cob_user)`, but that's extremely out of scope here. 😛

**Comment by n1zyy** [comment]
Demoed to Ann-Marie. Tagging as ready to merge!
